File: src/dotnet/Ansight.Tools.Preferences/PreferencesOptionsBuilderExtensions.cs
namespace Ansight.Tools.Preferences;

using System;

public static class PreferencesOptionsBuilderExtensions
{
    public static Options.OptionsBuilder WithPreferencesTools(this Options.OptionsBuilder builder)
        => builder.WithPreferencesTools(static _ => { });

    public static Options.OptionsBuilder WithPreferencesTools(
        this Options.OptionsBuilder builder,
        Action<PreferencesToolsOptionsBuilder> configure)
    {
        ArgumentNullException.ThrowIfNull(builder);
        ArgumentNullException.ThrowIfNull(configure);

        var optionsBuilder = PreferencesToolsOptions.CreateBuilder();
        configure(optionsBuilder);
        var options = optionsBuilder.Build();

        return builder.AddTools(new ITool[]
        {
            new ListPreferenceKeysTool(options),
            new GetPreferenceValueTool(options),
            new SetPreferenceValueTool(options),
            new RemovePreferenceKeyTool(options)
        });
    }
}

File: src/android/ansight-tools-preferences/src/main/kotlin/ai/ansight/tools/preferences/PreferencesOptionsBuilderExtensions.kt
package ai.ansight.tools.preferences

import ai.ansight.runtime.AnsightOptionsBuilder

fun AnsightOptionsBuilder.withPreferencesTools(): AnsightOptionsBuilder {
    return addTools(AndroidPreferencesTools.create())
}

fun AnsightOptionsBuilder.withPreferencesTools(
    options: AndroidPreferencesToolsOptions,
): AnsightOptionsBuilder {
    return addTools(AndroidPreferencesTools.create(options))
}

fun AnsightOptionsBuilder.withPreferencesTools(
    configure: AndroidPreferencesToolsOptionsBuilder.() -> Unit,
): AnsightOptionsBuilder {
    val builder = AndroidPreferencesToolsOptions.createBuilder()
    builder.configure()
    return withPreferencesTools(builder.build())
}
